File: backend/app/schemas/book.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from app.schemas.unit import UnitTreeNode


class BookSummaryResponse(BaseModel):
    id: UUID
    title: str
    author: str | None
    filename: str
    total_pages: int
    toc_extracted: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class BookProgressSummary(BaseModel):
    total_units: int
    completed_units: int
    in_progress_units: int
    not_started_units: int
    completion_percentage: float


class BookDetailResponse(BaseModel):
    id: UUID
    title: str
    author: str | None
    filename: str
    total_pages: int
    toc_extracted: bool
    units: list[UnitTreeNode]
    created_at: datetime

    model_config = {"from_attributes": True}


class BookProgressResponse(BaseModel):
    book_id: UUID
    title: str
    progress: BookProgressSummary


class DashboardResponse(BaseModel):
    books: list[BookSummaryResponse]
    total_books: int

File: frontend/src/types/book.ts
export interface BookSummary {
  id: string
  title: string
  author: string | null
  filename: string
  total_pages: number
  toc_extracted: boolean
  created_at: string
}

export interface UnitTreeNode {
  id: string
  title: string
  level: number
  order_index: number
  start_page: number
  end_page: number
  status: 'not_started' | 'in_progress' | 'completed'
  is_processed: boolean
  children: UnitTreeNode[]
}

export interface BookDetail {
  id: string
  title: string
  author: string | null
  filename: string
  total_pages: number
  toc_extracted: boolean
  units: UnitTreeNode[]
  created_at: string
}

export interface BookProgress {
  book_id: string
  title: string
  progress: {
    total_units: number
    completed_units: number
    in_progress_units: number
    not_started_units: number
    completion_percentage: number
  }
}

export interface Dashboard {
  books: BookSummary[]
  total_books: number
}
